### PR TITLE
Introduction of ostree-compliance-mode.spec file

### PR DESCRIPTION
**Pull request by iamianmullins**
Introduction of ostree-compliance-mode.spec file

Also add validation to fix some complier warnings

### CODE REVIEW

File: ostree-compliance-mode.spec [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+Version:        0.1
+Release:        1%{?dist}
+Summary:        ostree-compliance-mode
+
+License:        GPLv2
+URL:            https://github.com/ericcurtin/%{name}
+Source0:        %{url}/archive/main.tar.gz
+
+BuildRequires: meson
+BuildRequires: gcc-c++
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: This is a C codebase rather than a C++ one, so just gcc would be good here

File: ostree-compliance-mode.spec [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+%check
+%meson_test
+
+%install
+%meson_install
+
+%files
+%license LICENSE
+%doc README.md
+%{_bindir}/ostree-compliance-mode
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec:33]: You'll need ostree-compliance-mode-helper also, both with the correct permissions.

Might need to autogenerate /etc/ostree-compliance-mode.conf also at install time with the current rpm-ostree/ostree ref.

File: ostree-compliance-mode.spec [hunk 3]
```diff
@@ -0,0 +1,37 @@
+Name:           ostree-compliance-mode
+Version:        0.1
+Release:        1%{?dist}
+Summary:        ostree-compliance-mode
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: We need to write a summary;

`ostree-compliance-mode is a tool designed to purge protected data, leaving the system in a modifyable state in compliance with GPLv3.`

File: ostree-compliance-mode.spec [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+License:        GPLv2
+URL:            https://github.com/ericcurtin/%{name}
+Source0:        %{url}/archive/main.tar.gz
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: This should be similar to twincam, I made a real release here:

https://github.com/ericcurtin/ostree-compliance-mode/archive/refs/tags/0.1.tar.gz

**Comment by ericcurtin** [comment]
We should add something at install time along the lines of:

```
currently_deployed_ref=$(rpm-ostree status --json | python -c "import json,sys; print(json.load(sys.stdin)[\"deployments\"][$i][\"origin\"])")
echo "$currently_deployed_ref" > /etc/ostree-compliance-mode.conf
```

(but has to be tested, it's not gonna work unless ostree has something deployed). We must also add ostree, rpm-ostree and python as runtime dependencies to this spec (these are strictly required).

**Comment by ericcurtin** [comment]
We can leave the /etc/ostree-compliance-mode.conf stuff to v0.1.1 if needs be though, just to get something in the image.

**Comment by iamianmullins** [comment]
That sounds good, I'll apply some of that changes you suggested above, resubmit and then work on the other changes you've suggested for the next iteration of the spec file. Thanks

File: ostree-compliance-mode.c [hunk 5]
```diff
@@ -13,12 +13,15 @@
 #define PREFIX "/usr"
 #endif
 
-#define PERR(...)                \
-  do {                           \
-    char* str;                   \
-    asprintf(&str, __VA_ARGS__); \
-    perror(str);                 \
-    free(str);                   \
+#define PERR(...)                          \
```
Comment (ericcurtin) [inline ostree-compliance-mode.c:16]: Looks good

File: ostree-compliance-mode.spec [hunk 6]
```diff
@@ -0,0 +1,58 @@
+%global dracutdir %(pkg-config --variable=dracutdir dracut)
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: This rpm is not related to dracut, if you want some other spec files to reference, here's another resource:

https://gitlab.com/CentOS/automotive/rpms

Line 1 and 2 should be deleted here

Comment (iamianmullins) [reply ostree-compliance-mode.spec]: Agreed, that should have been removed. I'll take that out

File: ostree-compliance-mode.spec [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+Summary:        ostree-compliance-mode is a tool designed to purge protected data, leaving the system in a modifyable state in compliance with GPLv3
+
+License:        GPLv2
+URL:            https://github.com/ericcurtin/%{name}
+Source0:        %{url}/archive/%{version}/%{name}-%{version}.tar.gz
+
+BuildRequires: meson
+BuildRequires: gcc
+BuildRequires: ninja-build
+BuildRequires: ostree
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: The bottom 3 should be Requires as they are runtime requirements, like https://gitlab.com/CentOS/automotive/rpms/autosig-sample-staticdelta/-/blob/main/autosig-sample-staticdelta.spec file. We don't need `python3-gobject` though

File: ostree-compliance-mode.spec [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+BuildRequires: rpm-ostree
+BuildRequires: python3-devel
+
+%description
+%{summary}
+
+%prep
+%autosetup
+%global debug_package %{nil}
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: Do we need this?

`%global debug_package %{nil}`

File: ostree-compliance-mode.spec [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+%build
+%meson
+%meson_build
+
+%check
+%meson_test
+
+%install
+%meson_install
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: This `%install` section should be just two install commands with the correct permissions (one for ostree-compliance-mode, one for ostree-compliance-mode-helper), that puts the files in the correct directories. Look at various fedora/centos spec files for reference.

That's all that's required, we use install in rpm spec files, rather than cp, chmod, etc.

File: ostree-compliance-mode.spec (hunk 2 above)
Comment (ericcurtin) [reply ostree-compliance-mode.spec:33]: This section looks great!

File: ostree-compliance-mode.spec [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+%files
+%license LICENSE
+%doc README.md
+%{_bindir}/ostree-compliance-mode
+%{_sysconfdir}/ostree-compliance-mode.conf
+%{_prefix}/libexec/ostree-compliance-mode-helper
+
+%changelog
+* Thu Sep 8 2022 Ian Mullins<imullins@redhat.com> - 0.1-1
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: Should be:

`* Thu Sep 8 2022 Ian Mullins <imullins@redhat.com> - 0.1-1`

a space is missing between name and `<email>`

File: ostree-compliance-mode.spec (hunk 9 above)
Comment (ericcurtin) [reply ostree-compliance-mode.spec]: See `install --help`

File: ostree-compliance-mode.spec [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ninja -v -C build install
+chmod 4011 $RPM_BUILD_ROOT%{_bindir}/ostree-compliance-mode
+
+mkdir -p $RPM_BUILD_ROOT%{_prefix}/libexec
+cp ostree-compliance-mode-helper $RPM_BUILD_ROOT%{_prefix}/libexec/ostree-compliance-mode-helper
+chmod 100 $RPM_BUILD_ROOT%{_prefix}/libexec/ostree-compliance-mode-helper
+
+mkdir -p $RPM_BUILD_ROOT/etc
+currently_deployed_ref="fedora:fedora/35/aarch64/silverblue"
+echo "$currently_deployed_ref" > $RPM_BUILD_ROOT/etc/ostree-compliance-mode.conf
```
Comment (ericcurtin) [inline ostree-compliance-mode.spec]: We cannot hardcode things like `fedora/35/aarch64` here. What we can do in the %post section rather than the install section is determine the currently deployed ref and populate the `/etc/ostree-compliance-mode.conf` file. It could be aarch64, it could be x86_64, it could be fedora, it could be centos stream, etc.

It might be a good idea to leave population of `/etc/ostree-compliance-mode.conf` to a separate PR after this though. For version 0.1-2 maybe...

File: ostree-compliance-mode.spec (hunk 1 above)
Comment (ericcurtin) [reply ostree-compliance-mode.spec]: Looks great

File: ostree-compliance-mode.spec (hunk 4 above)
Comment (ericcurtin) [reply ostree-compliance-mode.spec]: LGTM

File: ostree-compliance-mode.spec (hunk 8 above)
Comment (iamianmullins) [reply ostree-compliance-mode.spec]: Without this, we seem to get the following build error. 

`RPM build errors:
    Empty %files file /builddir/build/BUILD/ostree-compliance-mode-0.1/debugsourcefiles.list
`

File: ostree-compliance-mode.spec (hunk 11 above)
Comment (iamianmullins) [reply ostree-compliance-mode.spec]: Yep, agreed, that was just temporarily hard-coded while I was still working on the other pieces. I've taken the ostree-compliance-mode.conf piece out for now as you suggested and will come back it. Thanks Eric!

**Comment by ericcurtin** [comment]
Partially merged this PR (the C part), to solve chicken and the egg problem. So this can build correctly. We now have a release 0.2

File: ostree-compliance-mode.spec (hunk 8 above)
Comment (ericcurtin) [reply ostree-compliance-mode.spec]: The problem with resolving it this way, is it's just silencing the error rather than fixing it, I believe using this disables generation of debuginfo.

**Comment by ericcurtin** [comment]
I made a couple of changes, it will be ok like this:

```
Name:          ostree-compliance-mode
Version:       0.2
Release:       1%{?dist}
Summary:       Leaves the system in a modifiable state in compliance with GPLv3

License:       GPLv2
URL:           https://github.com/ericcurtin/%{name}
Source0:       %{url}/archive/%{version}/%{name}-%{version}.tar.gz

BuildRequires: gcc

Requires:      ostree
Requires:      rpm-ostree
Requires:      python3

%description
%{summary}.

%prep
%autosetup

%build
${CC} ${RPM_OPT_FLAGS} -Wall -Winvalid-pch -Wextra -Wpedantic -Werror -std=gnu11 -Wstrict-aliasing -Wchar-subscripts -Wformat-security -Wmissing-declarations -Wpointer-arith -Wshadow -Wsign-compare -Wtype-limits -Wunused-function ostree-compliance-mode.c -o ostree-compliance-mode

%install
install -Dm 4411 ostree-compliance-mode %{buildroot}%{_bindir}/ostree-compliance-mode
install -Dm 500 %{SOURCE0} %{buildroot}%{_libexecdir}/ostree-compliance-mode-helper

%files
%license LICENSE
%doc README.md
%{_bindir}/ostree-compliance-mode
%{_libexecdir}/ostree-compliance-mode-helper

%changelog
* Mon Sep 12 2022 Ian Mullins <imullins@redhat.com> - 0.2-1
- Initial version
```

install -D, creates the directories if they don't exist. rpmlint let me know the descriptions were too long, so shortened them. Just used C compiler, since it's a one file project at the moment, meson is of little use and this way it was easier to fix the debug issue. Removed meson as a dependency.

**Comment by iamianmullins** [comment]
Looks great, thanks for the explanation of the changes you made, much appreciated!

**Comment by ericcurtin** [comment]
Rebased, squashed, and cleaned up log message and merged as fee0e96d88fa4c249392726b4604fc4461b044f0

```
git rebase origin/main # rebased
git rebase -i HEAD~2 # squashed 2 commits into one
```
